latch/type_engine/__init__.py

```diff
@@ -0,0 +1,393 @@
+        # class Foo
+        #   a: int
+        #   b: List[str]
+        #
+        # >>  Foo(a=1, b=["bar"]).schema()
+        #
+        # Member fields use the typing outlined [here](
+        # https://json-schema.org/understanding-json-schema/reference/type.html)
+        # and python types are inferred directly from these,
+        # skipping IDL representations completely.
+        #
+        # The one exception is @maximsmol's implementation of
+        # Union with (Dataclass JSON) python class fields, which
+        # will show *IDL* representations of Union variants in the
+        # "const" schema property:
+
+        # "oneOf": [
+        #   { "properties":
+        #     { "type":
+        #       { "const":
+        #         ...FlyteLiteralType JSON
+        #       }
+        #     }
+        #   },
+        # ]
+        #
+        # Logic used to handle Union fields is different from all
+        # other field types.
+
+        variants = []
+        for x in property_val["oneOf"]:
+            literal_type = gpjson.ParseDict(
+                x["properties"]["type"]["const"], _LiteralType()
+            )
+            variants.append(
+                guess_python_type(
+                    LiteralType.from_flyte_idl(literal_type),
+                )
+            )
+
+        return typing.Union[tuple(variants)]
+
+    property_type = property_val["type"]
+    if property_type == "object":
+        if property_val.get("$ref"):
+            nested_name = _get_schema_name(property_val)
+            # LatchFile and LatchDir are serialized as type "object" in JSON
+            # schema indistinguishable from eg. nested python classes.
+            if nested_name == "LatchfileSchema":
+                return LatchFile
+            elif nested_name == "LatchdirSchema":
+                return LatchDir
+            return _construct_class_from_json_schema(json_schema, nested_name)
+        elif property_val.get("additionalProperties"):
+            return typing.Dict[
+                str,
+                _guess_property_python_type(
+                    property_val["additionalProperties"], json_schema
+                ),
+            ]
+        else:
+            return typing.Dict[
+                str, _guess_property_python_type(property_val, json_schema)
+            ]
+    elif property_val["type"] == "array":
+        return typing.List[
+            _guess_property_python_type(property_val["items"], json_schema)
+        ]
+
+    elif property_type == "string":
+        return str
+    elif property_type == "integer":
+        return int
+    elif property_type == "boolean":
+        return bool
+    elif property_type == "number":
+        element_format = property_val["format"] if "format" in property_val else None
+        if element_format == "integer":
+            return int
+        else:
+            return float
+    return str
+
+
+def _construct_class_from_json_schema(json_schema: dict, schema_name: str):
```

Comment:
```suggestion
def _construct_class_from_json_schema(json_schema: dict[str, object], schema_name: str):
```
or you can use `TypedDict` and (maybe) avoid a lot of type checking pain